File: sdks/typescript/src/hook_event_values.ts
/** Hook event names and matcher semantics projected as local immutable values. */
export const HookEventCategory = Object.freeze({
  Tool: "Tool",
  Lifecycle: "Lifecycle",
  Subagent: "Subagent",
  Task: "Task",
  Error: "Error",
  Evolution: "Evolution",
} as const);

export type HookEventCategory = (typeof HookEventCategory)[keyof typeof HookEventCategory];

const EVENT_NAMES = [
  "PreToolUse",
  "PostToolUse",
  "PostToolUseFailure",
  "PermissionRequest",
  "PermissionDenied",
  "SessionStart",
  "SessionEnd",
  "Stop",
  "Notification",
  "UserPromptSubmit",
  "PreCompact",
  "PostCompact",
  "ConfigChange",
  "InstructionsLoaded",
  "PostToolBatch",
  "SubagentStart",
  "SubagentStop",
  "TaskCreated",
  "TaskStarted",
  "TaskCompleted",
  "StopFailure",
  "PluginLoaded",
  "PluginDisabled",
  "PostMemoryWrite",
  "MemoryLayerChange",
  "SkillCandidateDetected",
  "SkillLifecycleTransition",
  "SkillHealthCheck",
  "SkillPatchApplied",
  "SkillMergeApplied",
  "RulePromoted",
] as const;

export const HookEvent = Object.freeze({
  PreToolUse: "PreToolUse",
  PostToolUse: "PostToolUse",
  PostToolUseFailure: "PostToolUseFailure",
  PermissionRequest: "PermissionRequest",
  PermissionDenied: "PermissionDenied",
  SessionStart: "SessionStart",
  SessionEnd: "SessionEnd",
  Stop: "Stop",
  Notification: "Notification",
  UserPromptSubmit: "UserPromptSubmit",
  PreCompact: "PreCompact",
  PostCompact: "PostCompact",
  ConfigChange: "ConfigChange",
  InstructionsLoaded: "InstructionsLoaded",
  PostToolBatch: "PostToolBatch",
  SubagentStart: "SubagentStart",
  SubagentStop: "SubagentStop",
  TaskCreated: "TaskCreated",
  TaskStarted: "TaskStarted",
  TaskCompleted: "TaskCompleted",
  StopFailure: "StopFailure",
  PluginLoaded: "PluginLoaded",
  PluginDisabled: "PluginDisabled",
  PostMemoryWrite: "PostMemoryWrite",
  MemoryLayerChange: "MemoryLayerChange",
  SkillCandidateDetected: "SkillCandidateDetected",
  SkillLifecycleTransition: "SkillLifecycleTransition",
  SkillHealthCheck: "SkillHealthCheck",
  SkillPatchApplied: "SkillPatchApplied",
  SkillMergeApplied: "SkillMergeApplied",
  RulePromoted: "RulePromoted",
  ALL: EVENT_NAMES,
} as const);

export type HookEvent = (typeof HookEvent)[Exclude<keyof typeof HookEvent, "ALL">];

const TOOL_EVENTS: ReadonlySet<HookEvent> = new Set([
  HookEvent.PreToolUse,
  HookEvent.PostToolUse,
  HookEvent.PostToolUseFailure,
  HookEvent.PermissionRequest,
  HookEvent.PermissionDenied,
]);
const SUBAGENT_EVENTS: ReadonlySet<HookEvent> = new Set([
  HookEvent.SubagentStart,
  HookEvent.SubagentStop,
]);
const TASK_EVENTS: ReadonlySet<HookEvent> = new Set([
  HookEvent.TaskCreated,
  HookEvent.TaskStarted,
  HookEvent.TaskCompleted,
]);
const ERROR_EVENTS: ReadonlySet<HookEvent> = new Set([HookEvent.StopFailure]);
const EVOLUTION_EVENTS: ReadonlySet<HookEvent> = new Set([
  HookEvent.PostMemoryWrite,
  HookEvent.MemoryLayerChange,
  HookEvent.SkillCandidateDetected,
  HookEvent.SkillLifecycleTransition,
  HookEvent.SkillHealthCheck,
  HookEvent.SkillPatchApplied,
  HookEvent.SkillMergeApplied,
  HookEvent.RulePromoted,
]);

function assertEvent(event: string): asserts event is HookEvent {
  if (!(EVENT_NAMES as readonly string[]).includes(event)) {
    throw new TypeError(`invalid hook event: ${event}`);
  }
}

export function hookEventCategory(event: HookEvent): HookEventCategory {
  assertEvent(event);
  if (TOOL_EVENTS.has(event)) return HookEventCategory.Tool;
  if (SUBAGENT_EVENTS.has(event)) return HookEventCategory.Subagent;
  if (TASK_EVENTS.has(event)) return HookEventCategory.Task;
  if (ERROR_EVENTS.has(event)) return HookEventCategory.Error;
  if (EVOLUTION_EVENTS.has(event)) return HookEventCategory.Evolution;
  return HookEventCategory.Lifecycle;
}

export function hookEventAsStr(event: HookEvent): string {
  assertEvent(event);
  return event;
}

export function hookEventFromName(name: string): HookEvent | undefined {
  return (EVENT_NAMES as readonly string[]).includes(name) ? (name as HookEvent) : undefined;
}

export function hookEventIsToolEvent(event: HookEvent): boolean {
  return TOOL_EVENTS.has(event);
}

export function hookEventSupportsMatcher(event: HookEvent): boolean {
  assertEvent(event);
  return true;
}

File: sdks/python/src/echo_agent_sdk/hook_event_values.py
from __future__ import annotations

from enum import Enum
from typing import Final


class HookEventCategory(str, Enum):
    TOOL = "Tool"
    LIFECYCLE = "Lifecycle"
    SUBAGENT = "Subagent"
    TASK = "Task"
    ERROR = "Error"
    EVOLUTION = "Evolution"


class HookEvent(str, Enum):
    PRE_TOOL_USE = "PreToolUse"
    POST_TOOL_USE = "PostToolUse"
    POST_TOOL_USE_FAILURE = "PostToolUseFailure"
    PERMISSION_REQUEST = "PermissionRequest"
    PERMISSION_DENIED = "PermissionDenied"
    SESSION_START = "SessionStart"
    SESSION_END = "SessionEnd"
    STOP = "Stop"
    NOTIFICATION = "Notification"
    USER_PROMPT_SUBMIT = "UserPromptSubmit"
    PRE_COMPACT = "PreCompact"
    POST_COMPACT = "PostCompact"
    CONFIG_CHANGE = "ConfigChange"
    INSTRUCTIONS_LOADED = "InstructionsLoaded"
    POST_TOOL_BATCH = "PostToolBatch"
    SUBAGENT_START = "SubagentStart"
    SUBAGENT_STOP = "SubagentStop"
    TASK_CREATED = "TaskCreated"
    TASK_STARTED = "TaskStarted"
    TASK_COMPLETED = "TaskCompleted"
    STOP_FAILURE = "StopFailure"
    PLUGIN_LOADED = "PluginLoaded"
    PLUGIN_DISABLED = "PluginDisabled"
    POST_MEMORY_WRITE = "PostMemoryWrite"
    MEMORY_LAYER_CHANGE = "MemoryLayerChange"
    SKILL_CANDIDATE_DETECTED = "SkillCandidateDetected"
    SKILL_LIFECYCLE_TRANSITION = "SkillLifecycleTransition"
    SKILL_HEALTH_CHECK = "SkillHealthCheck"
    SKILL_PATCH_APPLIED = "SkillPatchApplied"
    SKILL_MERGE_APPLIED = "SkillMergeApplied"
    RULE_PROMOTED = "RulePromoted"

    def as_str(self) -> str:
        return self.value

    @classmethod
    def from_name(cls, name: str) -> HookEvent | None:
        try:
            return cls(name)
        except ValueError:
            return None

    def category(self) -> HookEventCategory:
        if self in _TOOL_EVENTS:
            return HookEventCategory.TOOL
        if self in _SUBAGENT_EVENTS:
            return HookEventCategory.SUBAGENT
        if self in _TASK_EVENTS:
            return HookEventCategory.TASK
        if self is HookEvent.STOP_FAILURE:
            return HookEventCategory.ERROR
        if self in _EVOLUTION_EVENTS:
            return HookEventCategory.EVOLUTION
        return HookEventCategory.LIFECYCLE

    def is_tool_event(self) -> bool:
        return self in _TOOL_EVENTS

    def supports_matcher(self) -> bool:
        return True


HOOK_EVENT_ALL: Final[tuple[HookEvent, ...]] = tuple(HookEvent)
_TOOL_EVENTS = frozenset(
    {
        HookEvent.PRE_TOOL_USE,
        HookEvent.POST_TOOL_USE,
        HookEvent.POST_TOOL_USE_FAILURE,
        HookEvent.PERMISSION_REQUEST,
        HookEvent.PERMISSION_DENIED,
    }
)
_SUBAGENT_EVENTS = frozenset({HookEvent.SUBAGENT_START, HookEvent.SUBAGENT_STOP})
_TASK_EVENTS = frozenset(
    {HookEvent.TASK_CREATED, HookEvent.TASK_STARTED, HookEvent.TASK_COMPLETED}
)
_EVOLUTION_EVENTS = frozenset(
    {
        HookEvent.POST_MEMORY_WRITE,
        HookEvent.MEMORY_LAYER_CHANGE,
        HookEvent.SKILL_CANDIDATE_DETECTED,
        HookEvent.SKILL_LIFECYCLE_TRANSITION,
        HookEvent.SKILL_HEALTH_CHECK,
        HookEvent.SKILL_PATCH_APPLIED,
        HookEvent.SKILL_MERGE_APPLIED,
        HookEvent.RULE_PROMOTED,
    }
)
